File: pulpogato-rest-tests/src/main/java/io/github/pulpogato/test/IgnoredTests.java
package io.github.pulpogato.test;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import lombok.Data;
import lombok.Getter;
import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.ObjectMapper;
import tools.jackson.dataformat.yaml.YAMLFactory;

/**
 * JUnit extension to ignore tests based on the version of the API and the schemaRef of the test.
 */
class IgnoredTests {
    private IgnoredTests() {
        // Empty Default Private Constructor. This should not be instantiated.
    }

    static Map<String, Map<String, String>> compute() {
        var tests = readTests();
        return tests.stream()
                .flatMap(ignoredTest -> ignoredTest.getVersions().stream()
                        .map(version ->
                                Map.entry(version, Map.entry(ignoredTest.getExample(), ignoredTest.getReason()))))
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        e -> Map.of(e.getValue().getKey(), e.getValue().getValue()),
                        (existing, replacement) -> {
                            Map<String, String> merged = new java.util.HashMap<>(existing);
                            merged.putAll(replacement);
                            return merged;
                        }));
    }

    private static List<IgnoredTest> readTests() {
        return new ObjectMapper(new YAMLFactory())
                .readValue(IgnoredTests.class.getResourceAsStream("/IgnoredTests.yml"), new TypeReference<>() {});
    }

    @Data
    static class IgnoredTest {
        private String example;
        private String reason;
        private List<String> versions;
    }

    /**
     * This maps the schemaRef of the test to the reason why the test is ignored.
     * Ideally, this should be a link to an issue on
     * <a href="https://github.com/github/rest-api-description/issues">GitHub/rest-api-description</a>.
     */
    @Getter
    private static final Map<String, Map<String, String>> causes = compute();
}

File: gradle/pulpogato-rest-codegen/src/main/kotlin/io/github/pulpogato/restcodegen/IgnoredTests.kt
package io.github.pulpogato.restcodegen

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory

object IgnoredTests {
    fun compute(): Map<String, Map<String, String>> {
        val tests =
            ObjectMapper(YAMLFactory()).readValue(
                IgnoredTests.javaClass.getResourceAsStream("/IgnoredTests.yml"),
                object : TypeReference<List<IgnoredTest>>() {},
            )
        val result =
            tests
                .flatMap { test ->
                    test.versions!!.map { ver -> mapOf("version" to ver, "example" to test.example!!, "reason" to test.reason) }
                }.groupBy { it["version"] }
                .map { (version, tests) -> version as String to tests.associate { it["example"] as String to it["reason"] as String } }
                .toMap()
        return result
    }

    class IgnoredTest() {
        var example: String? = null
        var reason: String? = null
        var versions: List<String>? = null
    }

    /**
     * This maps the schemaRef of the test to the reason why the test is ignored.
     * Ideally, this should be a link to an issue on [github/rest-api-description](https://github.com/github/rest-api-description/issues).
     */
    val causes = compute()
}